### PR TITLE
PERF: PeriodIndex speed up

### PR DESCRIPTION
**Pull request by max-sixty**
- [x] closes #https://github.com/pandas-dev/pandas/issues/14822
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master | flake8 --diff``
 - [x] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
great! pls add a whatsnew note. and if you can post the asv results would be great.

**Comment by MaximilianR** [comment]
```
    before     after       ratio
  [3ccb5013] [13aa1309]
+    4.48ms     6.10ms      1.36  period.period_standard_indexing.time_shape
-  659.85μs   572.45μs      0.87  period.period_standard_indexing.time_intersection
-  258.36μs   166.30μs      0.64  period.period_standard_indexing.time_series_loc
-  135.68μs    48.90μs      0.36  period.period_standard_indexing.time_get_loc
-    3.95ms    31.21μs      0.01  period.period_standard_indexing.time_shallow_copy
```

Not sure what's going with shape. It works fine in ipython:
```python
In [5]: index = pd.PeriodIndex(start=2000, periods=20000, freq='B')

In [7]: %timeit -n 1 -r 1 index.shape
1 loop, best of 1: 12.2 µs per loop
```

**Comment by jreback** [comment]
@MaximilianR that's fine. maybe needs np.random.seed to be set (to make it consitent). but no big deal.

File: asv_bench/benchmarks/period.py [hunk 1]
```diff
@@ -49,3 +49,28 @@ def time_value_counts_pindex(self):
         self.i.value_counts()
 
 
+class period_standard_indexing(object):
```
Comment (MaximilianR) [inline asv_bench/benchmarks/period.py]: Ideally this would be parameterized with Index type. For another day though.

**Comment by jreback** [comment]
lgtm. ping on green.

**Comment by jreback** [comment]
```
======================================================================
FAIL: test_getitem_ndim2 (pandas.tseries.tests.test_period.TestPeriodIndex)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Python27_64\envs\pandas\lib\site-packages\pandas\tseries\tests\test_period.py", line 2110, in test_getitem_ndim2
    self.assertEqual(result.shape, (len(idx), ))
AssertionError: Tuples differ: (3L, 1L) != (3,)
First tuple contains 1 additional elements.
First extra element 1:
1L
- (3L, 1L)
+ (3,)
------------------
```

**Comment by MaximilianR** [comment]
I'll look

**Comment by MaximilianR** [comment]
Any idea why this is here? Creating another axis on an index?

```
   2104     def test_getitem_ndim2(self):
   2105         idx = period_range('2007-01', periods=3, freq='M')
   2106
   2107         result = idx[:, None]
   2108         # MPL kludge, internally has incorrect shape
   2109         tm.assertIsInstance(result, PeriodIndex)
-> 2110         self.assertEqual(result.shape, (len(idx), ))
```

**Comment by MaximilianR** [comment]
Weird but different things happen with other indexes:

```python

In [7]: pd.DatetimeIndex(start='2000', periods=10, freq='B')
Out[7]:
DatetimeIndex(['2000-01-03', '2000-01-04', '2000-01-05', '2000-01-06',
               '2000-01-07', '2000-01-10', '2000-01-11', '2000-01-12',
               '2000-01-13', '2000-01-14'],
              dtype='datetime64[ns]', freq='B')

In [8]: pd.DatetimeIndex(start='2000', periods=10, freq='B')[:, None].shape
Out[8]: (10, 1)

In [9]: pd.DatetimeIndex(start='2000', periods=10, freq='B')[:, None]
Out[9]:
array([['2000-01-03T00:00:00.000000000'],
       ['2000-01-04T00:00:00.000000000'],
       ['2000-01-05T00:00:00.000000000'],
       ['2000-01-06T00:00:00.000000000'],
       ['2000-01-07T00:00:00.000000000'],
       ['2000-01-10T00:00:00.000000000'],
       ['2000-01-11T00:00:00.000000000'],
       ['2000-01-12T00:00:00.000000000'],
       ['2000-01-13T00:00:00.000000000'],
       ['2000-01-14T00:00:00.000000000']], dtype='datetime64[ns]')

In [10]: pd.DatetimeIndex(start='2000', periods=10, freq='B').shape
Out[10]: (10,)
```

**Comment by jreback** [comment]
this is a bit bogus, but I suspect this is in there for some matplotib compat (we pass periods to it I think for it to display some things).

**Comment by MaximilianR** [comment]
It's only changing the output of `.shape` - nothing else:

```python
In [3]: pd.PeriodIndex(start='2000', periods=10, freq='B')[:, None].shape
Out[3]: (10,)

In [4]: pd.PeriodIndex(start='2000', periods=10, freq='B')[:, None].values.shape
Out[4]: (10,)

In [5]: pd.PeriodIndex(start='2000', periods=10, freq='B')[:, None]._values.shape
Out[5]: (10, 1)
```

I hack change `.shape` to only look at the first dimension, or if nothing else fails are you OK taking that test out?

File: pandas/tseries/period.py [hunk 2]
```diff
@@ -305,7 +306,7 @@ def _simple_new(cls, values, name=None, freq=None, **kwargs):
             if (len(values) > 0 and is_float_dtype(values)):
                 raise TypeError("PeriodIndex can't take floats")
             else:
-                return PeriodIndex(values, name=name, freq=freq, **kwargs)
+                return cls(values, name=name, freq=freq, **kwargs)
```
Comment (MaximilianR) [inline pandas/tseries/period.py]: It's not what's causing the test failure (it could only be different if `cls != PeriodIndex`

**Comment by jreback** [comment]
ok taking the test out, if all plotting things pass (iow everything else is ok).

**Comment by jreback** [comment]
thanks!

**Pull request merged**